### PR TITLE
feat: adding temporal functions

### PR DESCRIPTION
**Pull request by rok**
This adds new scalar functions to work with timestamp/date/time types.

### CODE REVIEW

File: extensions/functions_datetime.yaml [hunk 1]
```diff
@@ -1,36 +1,52 @@
...
         return: i64
       - args:
-          - options: [ YEAR, MONTH, DAY, SECOND ]
+          - options: [ IS_LEAP_YEAR ]
+            name: The part of the value to extract.
+            required: true
+          - value: timestamp
+        return: boolean
+      - args:
+          - options: [ YEAR, QUARTER, MONTH, DAY, DAY_OF_WEEK, DAY_OF_YEAR, ISO_WEEK, IS_LEAP_YEAR, HOUR, SECOND, SUBSECOND, MILLISECOND, MICROSECOND, NANOSECOND, IS_DST ]
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:20]: Can you add definitions and what common data tools use each to a comment somewhere? That will make it easier to evaluate whether things should be included.

File: extensions/functions_datetime.yaml [hunk 2]
```diff
@@ -1,36 +1,52 @@
...
             name: The part of the value to extract.
             required: true
           - value: timestamp_tz
         return: i64
       - args:
-          - options: [ YEAR, MONTH, DAY ]
+          - options: [ IS_LEAP_YEAR, IS_DST ]
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:26]: I don't know what IS_DST means in the context of a instance in time.

File: extensions/functions_datetime.yaml [hunk 3]
```diff
@@ -57,8 +73,7 @@ scalar_functions:
           - value: date
           - value: interval_day
         return: timestamp
-  -
-    name: "add_intervals"
+  - name: "add_intervals"
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:76]: Why did you reformat all of these? Let's not do reformatting changes like this. To be consistent, if we think they are important let's add to a linter check as part of precommit.

File: extensions/functions_datetime.yaml [hunk 4]
```diff
@@ -193,3 +203,66 @@ scalar_functions:
...
+          - options: [ TIMEZONE ]
+            name: Timezone of provided local time.
+            required: true
+          - value: timestamp
+        return: timestamp_tz
+  - name: "strptime"
+    description: Parse string into timestamp/date/time using provided format.
+    impls:
+      - args:
+          - options: [ FORMAT ]
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:219]: Same as above, this seems to be a misunderstanding of options.

File: extensions/functions_datetime.yaml [hunk 5]
```diff
@@ -193,3 +203,66 @@ scalar_functions:
...
+    description: Convert local time to UTC using given local time's timezone.
+    impls:
+      - args:
+          - options: [ TIMEZONE ]
+            name: Timezone of provided local time.
+            required: true
+          - value: timestamp
+        return: timestamp_tz
+  - name: "strptime"
+    description: Parse string into timestamp/date/time using provided format.
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:216]: We need to declare the format rules this function uses.

File: extensions/functions_datetime.yaml [hunk 6]
```diff
@@ -193,3 +203,66 @@ scalar_functions:
...
+    impls:
+      - args:
+          - options: [ FORMAT ]
+            name: Expected format of the input timestamp/date/time.
+            required: true
+          - value: string
+        return: timestamp
+      - args:
+          - options: [ FORMAT ]
```
Comment (jacques-n) [inline extensions/functions_datetime.yaml:225]: This isn't allowed: two variations of the same function with the same input argument types. Both above and below are option + string.

**Review by jacques-n — CHANGES_REQUESTED** [review]
Several items need to be addressed.

**Comment by rok** [comment]
@jacques-n apologies I got distracted. Will address Monday.

File: extensions/functions_datetime.yaml (hunk 4 above)
Comment (rok) [reply extensions/functions_datetime.yaml:219]: Most likely. If options are enums, how do I bring in a string parameter?

File: extensions/functions_datetime.yaml (hunk 5 above)
Comment (rok) [reply extensions/functions_datetime.yaml:216]: Would that go into description here? I would start by adopting https://man7.org/linux/man-pages/man3/strptime.3.html

File: extensions/functions_datetime.yaml (hunk 6 above)
Comment (rok) [reply extensions/functions_datetime.yaml:225]: Removing.

File: extensions/functions_datetime.yaml [hunk 7]
```diff
@@ -193,3 +203,66 @@ scalar_functions:
...
+            name: Expected format of the input timestamp/date/time.
+            required: true
+          - value: string
+        return: time
+  - name: "strftime"
+    description: Convert timestamp/date/time to string using provided format
+    impls:
+      - args:
+          - options: [ FORMAT ]
```
[the remark this answers is not in the document]
Comment (rok) [reply extensions/functions_datetime.yaml:246]: Agreed. Will do.

**Comment by jacques-n** [comment]
Closing this due to inactivity. Please reopen if you want to proceed.

**Comment by rok** [comment]
@jacques-n I have pushed some changes, but I can't reopen the PR (permissions). Can you please reopen and review?

**Comment by jacques-n** [comment]
It looks like I can't reopen after you force pushed. 
image

I suggest just opening a new pr and referring to this one (or you could try force-pushing your old changes back, reopening, then overpushing). I'm not sure why Github makes this complicated...

**Comment by rok** [comment]
@jacques-n switching to #272

**Pull request closed without merging**